src/storage/src/hummock/store/version.rs

```diff
@@ -296,62 +286,33 @@ impl HummockReadVersion {
+                        .filter_map(|data| data.to_imm().map(|imm| imm.batch_id()))
+                        .collect();
 
                     // intersected batch_id order from oldest to newest
-                    let intersect_imm_ids = staging_sst_ref
+                    let intersect_imm_ids: HashSet<u64> = staging_sst_ref
                         .imm_ids
                         .iter()
                         .rev()
                         .copied()
                         .filter(|id| staging_imm_ids_from_imms.contains(id))
-                        .collect_vec();
+                        .collect();
 
                     if !intersect_imm_ids.is_empty() {
                         // Check 2)
-                        debug_assert!(check_subset_preserve_order(
-                            intersect_imm_ids.iter().copied(),
-                            self.staging.imm.iter().map(|imm| imm.batch_id()).rev(),
-                        ));
-
-                        // Check 3) and replace imms with a staging sst
-                        for imm_id in &intersect_imm_ids {
-                            if let Some(imm) = self.staging.imm.back() {
-                                if *imm_id == imm.batch_id() {
-                                    self.staging.imm.pop_back();
-                                }
-                            } else {
-                                let local_imm_ids = self
-                                    .staging
-                                    .imm
-                                    .iter()
-                                    .map(|imm| imm.batch_id())
-                                    .collect_vec();
-
-                                unreachable!(
-                                    "should not reach here staging_sst.size {},
-                                    staging_sst.imm_ids {:?},
-                                    staging_sst.epochs {:?},
-                                    local_imm_ids {:?},
-                                    intersect_imm_ids {:?}",
-                                    staging_sst_ref.imm_size,
-                                    staging_sst_ref.imm_ids,
-                                    staging_sst_ref.epochs,
-                                    local_imm_ids,
-                                    intersect_imm_ids,
-                                );
+                        self.staging.data.retain(|data| match data {
+                            StagingData::ImmMem(imm) => {
+                                !intersect_imm_ids.contains(&imm.batch_id())
                             }
-                        }
-                        self.staging.sst.push_front(staging_sst_ref);
+                            StagingData::Sst(_) => true,
+                        });
+                        self.staging.data.push(StagingData::Sst(staging_sst_ref));
+                        self.staging.data.sort_by_key(|data| data.min_epoch());
```

Comment:
Hmm... Here we remove the intersecting imms, push the new staging sst and sort the whole staging data by min_epoch. Though it will work, we actually have a stronger assumption here: staging sst should replace IMMs in consecutively.

```
// IMM(imm_id), SST(vec![imm_ids])
staging.data = [IMM(1), IMM(2), IMM(3), IMM(4)]

// Should be allowed in `HummockReadVersion::update`
Valid staging sst: SST([2,3])

// Should not be allowed in `HummockReadVersion::update` but allowed in this PR
Invalid staging sst: SST([1, 4])
```

I think we should not rely on `min_epoch` to sort the data after push but perform a point insert to replace concecutive IMMs with an assertion.

Reply:
Good catch, I will add this assert.